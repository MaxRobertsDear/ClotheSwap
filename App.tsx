import React, { useState } from 'react'
import { createStore, combineReducers, applyMiddleware } from 'redux'
import { Provider } from 'react-redux'
import { AppLoading } from 'expo'
import * as Font from 'expo-font'
import ReduxThunk from 'redux-thunk'
import * as firebase from 'firebase'

import { firebaseConfig } from './api-config'
// @ts-ignore
import productsReducer from './store/reducers/products'
// @ts-ignore
import cartReducer from './store/reducers/cart'
// @ts-ignore
import ordersReducer from './store/reducers/order'
import authReducer from './store/reducers/auth'
// @ts-ignore

import AppNavigator from './navigation/AppNavigator'

const fetchFonts = () => {
  return Font.loadAsync({
    'open-sans-regular': require('./assets/fonts/OpenSans-Regular.ttf'),
    'open-sans-bold': require('./assets/fonts/OpenSans-Bold.ttf'),
  })
}

firebase.initializeApp(firebaseConfig)

export const rootReducer = combineReducers({
  products: productsReducer,
  cart: cartReducer,
  orders: ordersReducer,
  auth: authReducer,
})

export type RootState = ReturnType<typeof rootReducer>
const store = createStore(rootReducer, applyMiddleware(ReduxThunk))

export default function App() {
  const [fontLoaded, setFontLoaded] = useState(false)
  if (!fontLoaded) {
    return (
      <AppLoading
        startAsync={fetchFonts}
        onFinish={() => {
          setFontLoaded(true)
        }}
      />
    )
  }

  return (
    <Provider store={store}>
      <AppNavigator />
    </Provider>
  )
}
