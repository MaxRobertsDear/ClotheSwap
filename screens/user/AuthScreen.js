import React, { useState, useEffect, useReducer, useCallback } from 'react'
import {
  Button,
  View,
  KeyboardAvoidingView,
  StyleSheet,
  ActivityIndicator,
  Alert,
  Platform,
} from 'react-native'
import { LinearGradient } from 'expo-linear-gradient'
import { useDispatch } from 'react-redux'

import Colors from '../../constants/Colors'
import Input from '../../components/UI/Input'
import Card from '../../components/UI/Card'
import * as authActions from '../../store/actions/auth'

const FORM_INPUT_UPDATE = 'FORM_INPUT_UPDATE'

const formReducer = (state, action) => {
  if (action.type === FORM_INPUT_UPDATE) {
    const updatedValues = {
      ...state.inputValues,
      [action.input]: action.value,
    }
    const updatedValidities = {
      ...state.inputValidities,
      [action.input]: action.isValid,
    }
    let updatedFormIsValid = true
    for (const key in updatedValidities) {
      updatedFormIsValid = updatedFormIsValid && updatedValidities[key]
    }
    return {
      formIsValid: updatedFormIsValid,
      inputValidities: updatedValidities,
      inputValues: updatedValues,
    }
  }
  return state
}

const AuthScreen = () => {
  const dispatch = useDispatch()
  const [isSignup, setIsSignup] = useState(false)
  const [isLoading, setIsLoading] = useState(false)
  const [error, setError] = useState()

  const [formState, dispatchFormState] = useReducer(formReducer, {
    inputValues: {
      email: '',
      password: '',
    },
    inputValidities: {
      email: true,
      password: true,
    },
    formIsValid: false,
  })

  useEffect(() => {
    if (error) {
      Alert.alert('An error occurred', error, [{ text: 'Okay' }])
    }
  }, [error])

  const signupLoginHandler = async () => {
    if (formState.formIsValid) {
      let action
      if (isSignup) {
        action = authActions.signup(
          formState.inputValues.email,
          formState.inputValues.password,
        )
      } else if (!isSignup) {
        action = authActions.login(
          formState.inputValues.email,
          formState.inputValues.password,
        )
      }
      setIsLoading(true)
      try {
        await dispatch(action)
      } catch (err) {
        setError(err.message)
        setIsLoading(false)
      }
    }
    return
  }

  const textChangeHandler = useCallback(
    (inputIdentifier, text) => {
      const emailRegex = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/
      let isValid = false
      if (inputIdentifier === 'password' && text.trim().length > 5) {
        isValid = true
      }
      if (inputIdentifier === 'email' && emailRegex.test(text.toLowerCase())) {
        isValid = true
      }
      dispatchFormState({
        type: FORM_INPUT_UPDATE,
        value: text,
        isValid: isValid,
        input: inputIdentifier,
      })
    },
    [dispatchFormState],
  )

  return (
    <KeyboardAvoidingView
      behavior={Platform.OS === 'ios' ? 'padding' : 'height'}
      style={styles.screen}
    >
      <LinearGradient colors={['#ffafbd', '#ffc3a0']} style={styles.gradient}>
        <Card style={styles.authContainer}>
          <Input
            label='Email'
            placeholder='example@example.com'
            keyboardType='email-address'
            autoCapitalize='none'
            returnKeyType='next'
            errorText={formState.inputValidities.email}
            value={formState.inputValues.email}
            onChangeText={(input) => textChangeHandler('email', input)}
          />
          <Input
            label='Password'
            placeholder='minimum 6 characters'
            keyboardType='default'
            autoCapitalize='none'
            secureTextEntry
            minLength={5}
            errorText={formState.inputValidities.password}
            value={formState.inputValues.password}
            onChangeText={(input) => textChangeHandler('password', input)}
          />
          <View style={styles.buttonContainer}>
            {isLoading ? (
              <ActivityIndicator size='small' color={Colors.primary} />
            ) : (
              <Button
                title={isSignup ? 'Sign Up' : 'Login'}
                color={Colors.primary}
                onPress={signupLoginHandler}
              />
            )}
            <Button
              title={`Switch to ${isSignup ? 'Login' : 'Sign Up'}`}
              color={Colors.accent}
              onPress={() => {
                setIsSignup((prevState) => !prevState)
              }}
            />
          </View>
        </Card>
      </LinearGradient>
    </KeyboardAvoidingView>
  )
}

const styles = StyleSheet.create({
  screen: {
    flex: 1,
  },
  gradient: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
  authContainer: {
    padding: 20,
    width: '80%',
    maxWidth: 400,
    maxHeight: 400,
  },
  buttonContainer: {
    marginTop: 10,
  },
})

export default AuthScreen
