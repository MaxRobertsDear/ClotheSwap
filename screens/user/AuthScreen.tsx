import React, { useState, useEffect, useReducer, useCallback } from 'react'
import {
  Button,
  View,
  KeyboardAvoidingView,
  StyleSheet,
  ActivityIndicator,
  Alert,
  Platform,
} from 'react-native'
import { LinearGradient } from 'expo-linear-gradient'
import { useDispatch } from 'react-redux'

import Colors from '../../constants/Colors'
import Input from '../../components/UI/Input'
import Card from '../../components/UI/Card'
import * as authActions from '../../store/actions/auth'

const FORM_INPUT_UPDATE = 'FORM_INPUT_UPDATE'

// interface iInputValues {
//   email: string;
//   password: string;
// }

interface iInputValidities {
  [key: string]: boolean;
}

const formReducer = (
  state: {
    // inputValues: iInputValues,
    inputValidities: iInputValidities,
    formIsValid: boolean,
  },
  action: { type: string, input: string, isValid: boolean },
) => {
  if (action.type === FORM_INPUT_UPDATE) {
    // const updatedValues = {
    //   ...state.inputValues,
    //   [action.input]: action.value,
    // }
    const updatedValidities = {
      ...state.inputValidities,
      [action.input]: action.isValid,
    }
    let updatedFormIsValid = true
    for (const key in updatedValidities) {
      updatedFormIsValid = updatedFormIsValid && updatedValidities[key]
    }
    return {
      formIsValid: updatedFormIsValid,
      inputValidities: updatedValidities,
      // inputValues: updatedValues,
    }
  }
  return state
}

const AuthScreen = () => {
  const dispatch = useDispatch()
  const [isSignup, setIsSignup] = useState(false)
  const [isLoading, setIsLoading] = useState(false)
  const [error, setError] = useState()
  const [email, setEmail] = useState('')
  const [password, setPassword] = useState('')

  const [formState, dispatchFormState] = useReducer(formReducer, {
    // inputValues: {
    //   email: '',
    //   password: '',
    // },
    inputValidities: {
      email: true,
      password: true,
    },
    formIsValid: false,
  })

  useEffect(() => {
    if (error) {
      Alert.alert('An error occurred', error, [{ text: 'Okay' }])
    }
  }, [error])

  const signupLoginHandler = async () => {
    if (formState.formIsValid) {
      let action
      if (isSignup) {
        action = authActions.signup({
          email: email,
          password: password,
        })
      } else if (!isSignup) {
        action = authActions.login({
          email: email,
          password: password,
        })
      }
      setIsLoading(true)
      try {
        await dispatch(action)
      } catch (err) {
        setError(err.message)
        setIsLoading(false)
      }
    }
    return
  }

  const formFieldValidator = useCallback(
    (inputIdentifier, text) => {
      const emailRegex = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/
      let isValid = false
      if (inputIdentifier === 'password' && text.trim().length > 5) {
        isValid = true
      }
      if (inputIdentifier === 'email' && emailRegex.test(text.toLowerCase())) {
        isValid = true
      }
      dispatchFormState({
        type: FORM_INPUT_UPDATE,
        // value: text,
        isValid: isValid,
        input: inputIdentifier,
      })
    },
    [dispatchFormState],
  )

  return (
    <KeyboardAvoidingView
      behavior={Platform.OS === 'ios' ? 'padding' : 'height'}
      style={styles.screen}
    >
      <LinearGradient colors={['#ffafbd', '#ffc3a0']} style={styles.gradient}>
        <Card style={styles.authContainer}>
          <Input
            label='Email'
            placeholder='example@example.com'
            keyboardType='email-address'
            autoCapitalize='none'
            returnKeyType='next'
            errorText={formState.inputValidities.email}
            value={email}
            onChangeText={(input: string) => setEmail(input)}
            onBlur={() => formFieldValidator('email', email)}
          />
          <Input
            label='Password'
            placeholder='minimum 6 characters'
            keyboardType='default'
            autoCapitalize='none'
            secureTextEntry
            minLength={5}
            errorText={formState.inputValidities.password}
            value={password}
            onChangeText={(input: string) => setPassword(input)}
            onBlur={() => formFieldValidator('password', password)}
          />
          <View style={styles.buttonContainer}>
            {isLoading ? (
              <ActivityIndicator size='small' color={Colors.primary} />
            ) : (
              <Button
                title={isSignup ? 'Sign Up' : 'Login'}
                color={Colors.primary}
                onPress={signupLoginHandler}
              />
            )}
            <Button
              title={`Switch to ${isSignup ? 'Login' : 'Sign Up'}`}
              color={Colors.accent}
              onPress={() => {
                setIsSignup((prevState) => !prevState)
              }}
            />
          </View>
        </Card>
      </LinearGradient>
    </KeyboardAvoidingView>
  )
}

const styles = StyleSheet.create({
  screen: {
    flex: 1,
  },
  gradient: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
  authContainer: {
    padding: 20,
    width: '80%',
    maxWidth: 400,
    maxHeight: 400,
    shadowColor: 'black',
    shadowOpacity: 0.27,
    shadowOffset: { width: 0, height: 2 },
    shadowRadius: 8,
  },
  buttonContainer: {
    marginTop: 10,
  },
})

export default AuthScreen
