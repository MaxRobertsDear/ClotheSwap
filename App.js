import React, { useState, createContext } from 'react'
import { Text, View, Platform } from 'react-native'
import { NavigationContainer } from '@react-navigation/native'
import { createStackNavigator } from '@react-navigation/stack'
import { createDrawerNavigator } from '@react-navigation/drawer'
import {
  createStore,
  combineReducers,
  Middleware,
  applyMiddleware,
} from 'redux'
import { Provider } from 'react-redux'
import { AppLoading } from 'expo'
import * as Font from 'expo-font'
import ReduxThunk from 'redux-thunk'

import productsReducer from './store/reducers/products'
import cartReducer from './store/reducers/cart'
import ordersReducer from './store/reducers/order'
import ProductsOverviewScreen from './screens/shop/ProductsOverviewScreen'
import ProductDetailScreen from './screens/shop/ProductDetailScreen'
import Colors from './constants/Colors'
import CartScreen from './screens/shop/CartScreen'
import OrdersScreen from './screens/shop/OrdersScreen'
import UserProductsScreen from './screens/user/UserProductsScreen'
import EditProductsScreen from './screens/user/EditProductScreen'
import AuthScreen from './screens/user/AuthScreen'

const fetchFonts = () => {
  return Font.loadAsync({
    'open-sans-regular': require('./assets/fonts/OpenSans-Regular.ttf'),
    'open-sans-bold': require('./assets/fonts/OpenSans-Bold.ttf'),
  })
}

export const rootReducer = combineReducers({
  products: productsReducer,
  cart: cartReducer,
  orders: ordersReducer,
})
const store = createStore(rootReducer, applyMiddleware(ReduxThunk))

const Stack = createStackNavigator()
const Drawer = createDrawerNavigator()

export default function App() {
  const [fontLoaded, setFontLoaded] = useState(false)
  const [isLoggedIn, setIsLoggedIn] = useState(false)

  if (!fontLoaded) {
    return (
      <AppLoading
        startAsync={fetchFonts}
        onFinish={() => {
          setFontLoaded(true)
        }}
      />
    )
  }

  const Home = () => {
    return (
      <Stack.Navigator>
        <Stack.Screen
          name='ProductsOverview'
          component={ProductsOverviewScreen}
          options={{
            title: 'Products Overview',
            headerStyle: {
              backgroundColor:
                Platform.OS === 'android' ? Colors.primary : 'white',
            },
            headerTitleStyle: {
              color: Platform.OS === 'android' ? 'whitea' : Colors.primary,
            },
          }}
        />
        <Stack.Screen
          name='ProductDetails'
          component={ProductDetailScreen}
          options={{
            title: 'Product Details',
          }}
        />
        <Stack.Screen
          name='CartScreen'
          component={CartScreen}
          options={{
            title: 'Cart',
          }}
        />
      </Stack.Navigator>
    )
  }

  const Admin = () => {
    return (
      <Stack.Navigator>
        <Stack.Screen
          name='UserProductsScreen'
          component={UserProductsScreen}
          options={{
            title: 'Your Products',
            headerStyle: {
              backgroundColor:
                Platform.OS === 'android' ? Colors.primary : 'white',
            },
            headerTitleStyle: {
              color: Platform.OS === 'android' ? 'white' : Colors.primary,
            },
          }}
        />
        <Stack.Screen
          name='EditProductsScreen'
          component={EditProductsScreen}
          options={{
            title: 'Edit Your Products',
            headerStyle: {
              backgroundColor:
                Platform.OS === 'android' ? Colors.primary : 'white',
            },
            headerTitleStyle: {
              color: Platform.OS === 'android' ? 'white' : Colors.primary,
            },
          }}
        />
      </Stack.Navigator>
    )
  }

  const Orders = () => {
    return (
      <Stack.Navigator>
        <Stack.Screen
          name='OrdersScreen'
          component={OrdersScreen}
          options={{
            title: 'Orders',
            headerStyle: {
              backgroundColor:
                Platform.OS === 'android' ? Colors.primary : 'white',
            },
            headerTitleStyle: {
              color: Platform.OS === 'android' ? 'whitea' : Colors.primary,
            },
          }}
        />
      </Stack.Navigator>
    )
  }

  let screenStack
  console.log('-------------------------------------', isLoggedIn)
  if (isLoggedIn) {
    screenStack = (
      <Drawer.Navigator initialRouteName='Home' drawerType='slide'>
        <Drawer.Screen name='Home' component={Home} />
        <Drawer.Screen name='Orders' component={Orders} />
        <Drawer.Screen name='Admin' component={Admin} />
      </Drawer.Navigator>
    )
  } else {
    screenStack = (
      <Stack.Navigator>
        <Stack.Screen
          name='AuthScreen'
          component={AuthScreen}
          initialParams={isLoggedIn}
          options={{
            title: 'Auth Screen',
            headerStyle: {
              backgroundColor:
                Platform.OS === 'android' ? Colors.primary : 'white',
            },
            headerTitleStyle: {
              color: Platform.OS === 'android' ? 'whitea' : Colors.primary,
            },
          }}
        />
        <Stack.Screen name='Home' component={Home} />
      </Stack.Navigator>
    )
  }

  return (
    <Provider store={store}>
      <NavigationContainer>{screenStack}</NavigationContainer>
    </Provider>
  )
}
