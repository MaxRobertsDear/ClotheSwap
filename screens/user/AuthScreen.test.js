import React from 'react'
import AuthScreen from './AuthScreen'
import { fireEvent, render, wait } from '@testing-library/react-native'
import { createStore } from 'redux'
import { Provider } from 'react-redux'
import { rootReducer } from '../../App'
import renderer from 'react-test-renderer'

function renderWithRedux(
  ui,
  { initialState, store = createStore(rootReducer, initialState) } = {},
) {
  return {
    ...renderer.create(<Provider store={store}>{ui}</Provider>).toJSON(),
    store,
  }
}
// test('renders correctly', async () => {
// 	const { container, debug, queryByText } = renderWithRedux(<AuthScreen />);
// 	// debug();
// 	// expect(queryByText('Password')).toBeTruthy();
// 	expect(container).toMatchSnapshot();
// });
test('renders correctly', async () => {
  const tree = renderWithRedux(<AuthScreen />)
  expect(tree).toMatchSnapshot()
})
