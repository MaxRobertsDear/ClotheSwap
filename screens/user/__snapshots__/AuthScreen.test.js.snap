// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Object {
  "children": Array [
    <ViewManagerAdapter_ExpoLinearGradient
      proxiedProperties={
        Object {
          "colors": Array [
            4294946749,
            4294951840,
          ],
          "endPoint": undefined,
          "locations": undefined,
          "startPoint": undefined,
        }
      }
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "white",
            "borderRadius": 10,
            "elevation": 5,
            "maxHeight": 400,
            "maxWidth": 400,
            "padding": 20,
            "shadowColor": "black",
            "shadowOffset": Object {
              "height": 2,
              "width": 0,
            },
            "shadowOpacity": 0.27,
            "shadowRadius": 8,
            "width": "80%",
          }
        }
      >
        <View
          style={
            Object {
              "width": "100%",
            }
          }
        >
          <Text
            style={
              Object {
                "fontFamily": "open-sans-bold",
                "marginVertical": 8,
              }
            }
          >
            Email
          </Text>
          <TextInput
            allowFontScaling={true}
            autoCapitalize="none"
            errorText={false}
            keyboardType="email-address"
            label="Email"
            onChangeText={[Function]}
            rejectResponderTermination={true}
            returnKeyType="next"
            style={
              Object {
                "borderBottomColor": "#ccc",
                "borderBottomWidth": 1,
                "paddingHorizontal": 2,
                "paddingVertical": 5,
              }
            }
            underlineColorAndroid="transparent"
            value=""
          />
          <Text>
            Please enter a valid 
            Email
          </Text>
        </View>
        <View
          style={
            Object {
              "width": "100%",
            }
          }
        >
          <Text
            style={
              Object {
                "fontFamily": "open-sans-bold",
                "marginVertical": 8,
              }
            }
          >
            Password
          </Text>
          <TextInput
            allowFontScaling={true}
            autoCapitalize="none"
            errorText={false}
            keyboardType="default"
            label="Password"
            minLength={5}
            onChangeText={[Function]}
            rejectResponderTermination={true}
            secureTextEntry={true}
            style={
              Object {
                "borderBottomColor": "#ccc",
                "borderBottomWidth": 1,
                "paddingHorizontal": 2,
                "paddingVertical": 5,
              }
            }
            underlineColorAndroid="transparent"
            value=""
          />
          <Text>
            Please enter a valid 
            Password
          </Text>
        </View>
        <View
          style={
            Object {
              "marginTop": 10,
            }
          }
        >
          <View
            accessibilityRole="button"
            accessibilityStates={Array []}
            accessible={true}
            focusable={true}
            isTVSelectable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <View
              style={
                Array [
                  Object {},
                ]
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#007AFF",
                      "fontSize": 18,
                      "padding": 8,
                      "textAlign": "center",
                    },
                    Object {
                      "color": "#C2185B",
                    },
                  ]
                }
              >
                Login
              </Text>
            </View>
          </View>
          <View
            accessibilityRole="button"
            accessibilityStates={Array []}
            accessible={true}
            focusable={true}
            isTVSelectable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <View
              style={
                Array [
                  Object {},
                ]
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#007AFF",
                      "fontSize": 18,
                      "padding": 8,
                      "textAlign": "center",
                    },
                    Object {
                      "color": "#FFC107",
                    },
                  ]
                }
              >
                Switch to Sign Up
              </Text>
            </View>
          </View>
        </View>
      </View>
    </ViewManagerAdapter_ExpoLinearGradient>,
  ],
  "props": Object {
    "onLayout": [Function],
    "style": Array [
      Object {
        "flex": 1,
      },
      Object {
        "paddingBottom": 0,
      },
    ],
  },
  "store": Object {
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
    Symbol(observable): [Function],
  },
  "type": "View",
}
`;
